components/identity-mgt/org.wso2.carbon.identity.mgt/src/main/java/org/wso2/carbon/identity/mgt/listener/IdentityUserIdResolverListener.java

```diff
@@ -0,0 +1,934 @@
+/*
+ * Copyright (c) 2019, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ * WSO2 Inc. licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.mgt.listener;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.core.AbstractIdentityUserOperationEventListener;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+import org.wso2.carbon.user.api.Permission;
+import org.wso2.carbon.user.core.UserStoreException;
+import org.wso2.carbon.user.core.UserStoreManager;
+import org.wso2.carbon.user.core.common.AbstractUserStoreManager;
+import org.wso2.carbon.user.core.common.AuthenticationResult;
+import org.wso2.carbon.user.core.common.User;
+import org.wso2.carbon.user.core.internal.UMListenerServiceComponent;
+import org.wso2.carbon.user.core.listener.UniqueIDUserOperationEventListener;
+import org.wso2.carbon.user.core.listener.UserOperationEventListener;
+import org.wso2.carbon.user.core.model.Condition;
+import org.wso2.carbon.user.core.model.UserClaimSearchEntry;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This is an implementation of UserOperationEventListener. This will call the
+ * relevant UniqueIDUserOperationEventListener methods when the user name related listeners are being called. This has
+ * been implemented to preserve the backwards compatibility.
+ */
+public class IdentityUserIdResolverListener extends AbstractIdentityUserOperationEventListener {
+
+    private static final Log log = LogFactory.getLog(IdentityUserIdResolverListener.class);
+
+    @Override
+    public int getExecutionOrderId() {
+        int orderId = getOrderId();
+        if (orderId != IdentityCoreConstants.EVENT_LISTENER_ORDER_ID) {
+            return orderId;
+        }
+        return 15;
+    }
+
+    @Override
+    public boolean doPreAuthenticate(String userName, Object credential, UserStoreManager userStoreManager)
+            throws UserStoreException {
+
+        if (!isEnable()) {
+            return true;
+        }
+
+        String userID = ((AbstractUserStoreManager) userStoreManager).getUserIDFromUserName(userName);
+        for (UserOperationEventListener listener : UMListenerServiceComponent.getUserOperationEventListeners()) {
```

Comment:
No need to trigger id listeners if the userID is null. That may cause to null pointers in listeners level